imap_processing/ultra/l1b/de.py

```diff
@@ -13,46 +30,172 @@ def calculate_de(de_dataset: xr.Dataset, name: str) -> xr.Dataset:
+    )
+    etof[ph_indices], xc[ph_indices] = get_coincidence_positions(
+        de_dataset.isel(epoch=ph_indices), t2[ph_indices], f"ultra{sensor}"
+    )
+    ctof[ph_indices] = get_ctof(tof[ph_indices], r[ph_indices], "PH")
+
+    # SSD
+    ssd_indices = np.nonzero(np.isin(de_dataset["STOP_TYPE"], StopType.SSD.value))[0]
+    tof[ssd_indices] = get_ssd_tof(de_dataset, xf)
+    yb[ssd_indices], _, ssd_number = get_ssd_back_position_and_tof_offset(de_dataset)
+    xc[ssd_indices] = np.zeros(len(ssd_indices))
+    xb[ssd_indices] = np.zeros(len(ssd_indices))
+    etof[ssd_indices] = np.zeros(len(ssd_indices))
+    d[ssd_indices], yf[ssd_indices] = get_front_y_position(
+        de_dataset["START_TYPE"].data[ssd_indices], yb[ssd_indices]
+    )
+    energy[ssd_indices] = get_energy_ssd(de_dataset, ssd_number)
+    r[ssd_indices] = get_path_length(
+        (xf[ssd_indices], yf[ssd_indices]),
+        (xb[ssd_indices], yb[ssd_indices]),
+        d[ssd_indices],
+    )
+    species_bin[ssd_indices] = determine_species_ssd(
+        energy[ssd_indices],
+        tof[ssd_indices],
+        r[ssd_indices],
+    )
+    ctof[ssd_indices] = get_ctof(tof[ssd_indices], r[ssd_indices], "SSD")
+
+    # Combine ph_yb and ssd_yb along with their indices
+    de_dict["x_front"] = xf.astype(np.float32)
+    de_dict["y_front"] = yf
+    de_dict["x_back"] = xb
+    de_dict["y_back"] = yb
+    de_dict["x_coin"] = xc
+    de_dict["tof_start_stop"] = tof
+    de_dict["tof_stop_coin"] = etof
+    de_dict["tof_corrected"] = ctof
+    de_dict["front_back_distance"] = d
+    de_dict["path_length"] = r
+
+    keys = [
+        "coincidence_type",
+        "start_type",
+        "event_type",
+        "de_event_met",
+    ]
+    dataset_keys = ["COIN_TYPE", "START_TYPE", "STOP_TYPE", "SHCOARSE"]
+
+    de_dict.update(
+        {key: de_dataset[dataset_key] for key, dataset_key in zip(keys, dataset_keys)}
+    )
+
+    vx_ultra, vy_ultra, vz_ultra = get_unit_vector(
+        (de_dict["x_front"], de_dict["y_front"]),
+        (de_dict["x_back"], de_dict["y_back"]),
+        de_dict["front_back_distance"],
+        de_dict["tof_start_stop"],
+    )
+
+    de_dict["vx_ultra"] = vx_ultra.astype(np.float32)
+    de_dict["vy_ultra"] = vy_ultra.astype(np.float32)
+    de_dict["vz_ultra"] = vz_ultra.astype(np.float32)
+    de_dict["energy"] = energy
+    de_dict["species"] = species_bin
+
+    # Annotated Events.
+    # TODO: since the pointing (dps) frame is not for this timerange this will not work.
+    # position = np.stack(
+    #     (de_dict["vx_ultra"], de_dict["vy_ultra"], de_dict["vz_ultra"]), axis=-1
+    # )
+    #
+    # ultra_frame = getattr(SpiceFrame, f"IMAP_ULTRA_{sensor}")
+    # sc_velocity, sc_dps_velocity, helio_velocity = get_annotated_particle_velocity(
+    #     de_dataset.data_vars["EVENTTIMES"],
+    #     position,
+    #     ultra_frame,
+    #     SpiceFrame.IMAP_DPS,
+    #     SpiceFrame.IMAP_SPACECRAFT,
+    # )
```

Comment:
If I understand this, the problem is that you don't have test data from the same time as the coverage provided by the simulated pointing frame that we have?
I would recommend fixing that issue in the test rather than in this code. You could mock out the `get_annotated_particle_velocity` function in the test and have it return the full NaN ndarrays that you have coded in below.

Reply:
I had never done that before. That came in handy. Thanks!